docs/Tutorial_DiscordMERLIN.ipynb

```diff
@@ -0,0 +1,1975 @@
+{
```

Comment:
<pre class="ql-syntax" spellcheck="false">Line #24.        P : numpy.ndarray
</pre><p>This doesn't make sense. You shouldn't be having to pass in some approximated <code>P</code> when you can compute it yourself inside the function. Instead, you should simply do <code>prescrump=False</code> as your parameter and, if <code>prescrump=True</code> then compute the approximated <code>P</code> using <code>_prescrump</code> instead of <code>prescrump</code> . Right now, you are needing to deal with whether or not you are using <code>T</code> or <code>ts</code> which should signal to you that something is really poorly designed here if you need to take so much care. We need to write the code in such a way that everything is obvious. Having to think about <code>T</code> vs <code>ts</code> is not obvious!</p><p> </p><p>More importantly, you need to start recognizing that the design of each function needs to be thoughtfully considered (i.e., what assumptions are you making, are those assumptions obvious/logical from the user's standpoint, when are you giving the user too much control, what can you make "easier" for the user, etc). This is one of those places where there is instant "code smell" as it is really, fishy. This docstring of <code>not after preprocessing)</code>  made it abundantly obvious that something was bad. I've mentioned this before and so you should really take note of places where you feel the need to add an extra comment or note as those are the places where maybe a better design choice would help. Sometimes, we don't have a choice but, most of the time, we do.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/505/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='33'/>

Reply:
<p>I am beginning to learn something new!... Even if the function is private, I must still care about user. In other words, although I know what I am doing in the outer wrapper and what I am going to pass to these (private) functions, I still need to make sure that each function alone makes sense (i.e. well-designed/organized) </p>

Reply:
I took a  look at `_prescrump()`, what a big module (IMHO?) !! I need sometime to go over it. 

> Right now, you are needing to deal with whether or not you are using `T` or `ts` which should signal to you that something is really poorly designed here if you need to take so much care.

Right. I wrongfully thought things are okay just because this is a private function (for private use) in `_murlin`.

Reply:
>I've mentioned this before and so you should really take note of places where you feel the need to add an extra comment or note as those are the places where maybe a better design choice would help. 

Sorry for this! I should learn faster...

Reply:
<p>Yes! In my mind, the difference between a private function and public function is that the private function has no guarantees and may be subject to change in the future (so use at your own risk). Whereas a public function is essentially agreed that the API should not change unless the "major version" is bumped up. However, in BOTH private and public functions, there will still be users using them and so we must always try to design them well! </p><p> </p><p>To be clear, for STUMPY, just because something is a public function it doesn't mean that it will get added to our API documentation page. A good example is the <code>core.apply_exclusion_zone</code>  function. It is technically a public function but it is "public" because, after so many years, the API is unlikely to change and so it is relatively stable. What did change is the underlying <code>core._apply_exclusion_zone</code>  function. It was a minor change that allowed us to accommodate a few more use cases but not drastically. When we design, there is a lot to consider and it may not be easy to "get it right" but we need to think about the "majority use case".</p>

Reply:
<blockquote>Sorry for this! I should learn faster...</blockquote><p>No need to apologize! I only made the comment to help connect the current situation to a point from before. There is a lot to keep track of so it's good to reinforce the lessons learned.</p><p> </p><p> </p><blockquote>I took a look at&nbsp;<code>_prescrump()</code>, what a big module (IMHO?) !! I need sometime to go over it.</blockquote><p>For the record, <code>_prescrump()</code>  was mostly borrowed from the original Matlab code. I don't think you really need to spend time understanding it as it may not be worth your time. But I leave that choice up to you. It took me forever to fully grasp what was going on when I had implemented it</p>

Reply:
> so it's good to reinforce the lessons learned

Yes! Please keep doing so (and I should definitely not rush into things 😄 ) 

> So, the temporary use of `P` will be forgotten and then nobody will ever change it.

Just trying to understand your mindset here... would you please elaborate what you meant by "temporary"? And, when you said "nobody will ever change it", do you mean...like improvement and making the algorithm/function better in the future?

---

> For the record, `_prescrump()` was mostly borrowed from the original Matlab code. I don't think you really need to spend time understanding it as it may not be worth your time.

My goal is just to understand "HOW" to use it. I am having hard time to understand how to use it. So, I took a look at `prescrump` (not `_prescrump`), and if I understand correctly, I should add the following block to `_murlin` (as you mentioned in another comment after moving `core.preprocess(...)` to the top of `_murlin(...)` ). Again, we are doing this to increase safety and avoid potential mistakes in the future. Correct?
```
    n_A = T_A.shape[0]
    n_B = T_B.shape[0]
    l = n_A - m + 1
    P_squared = np.full(l, np.inf, dtype=np.float64)
    I = np.full(l, -1, dtype=np.int64)
    squared_distance_profile = np.empty(n_B - m + 1, dtype=np.float64)

    if s is None:  # pragma: no cover
        s = excl_zone

    for i in np.random.permutation(range(0, l, s)).astype(np.int64):
        QT = core.sliding_dot_product(T_A[i : i + m], T_B)
        _prescrump(
            T_A,
            T_B,
            m,
            M_T,
            Σ_T,
            μ_Q,
            σ_Q,
            QT,
            i,
            s,
            squared_distance_profile,
            P_squared,
            I,
            excl_zone,
        )

    P = np.sqrt(P_squared)

```
(and `T_B`  is basically just `T_A` here).

Now, all sounds good. But, I am trying to see how I can `re-use` it in `_find_discords`. Because, the way you explained about setting `prescrump=False`, and then do calculation if True, doesn't that mean I am not taking advantage of already calculated `P`? How about doing `prescrump=None`. So, if `None`, we will calculate it (naively) inside `_find_discords` by using the same block of code above (I said "naively"; because, a better approach is to just to get approx `P` for `cands_idx`, for instance by concatenating them using `np.inf` spacer(?!))

Reply:
<p>&gt;Just trying to understand your mindset here... would you please elaborate what you meant by "temporary"? And, when you said "nobody will ever change it", do you mean...like improvement and making the algorithm/function better in the future?</p><p> </p><p>Yes, in my experience, when somebody is writing code and thinks, "Well, I'll just do this thing really quick and then, later when I have more time, I'll go back and fix it or make it better", almost 99% of the time that thing never "gets fixed later". So, what is often "temporary" at the time just gets neglected and then forgotten. Later, when other people come to look at that part of the code, they have no idea why it's there or even how to fix it. This is really bad and, as you'd imagine, the more times it happens the more of these things add up. So, it is always best to be proactive and spend the time upfront to make it as good as possible because, in the future, we won't have time or remember the details of how to fix it. A great example of this is imagine if you were to go back to look at the "geometric chains" code that you had written and then think about how much time it would take you to update/fix that code now. It would probably be extremely painful and dreaded. So, it's always best to put in the time now then to come back to it later. </p>

Reply:
> missed that you were using `_find_discords` within a `for-loop` inside of `_murlin` as I hadn't looked that far down.

What do you think? In your opinion, is it a poor (or NOT a good) design? (because you had to notice another thing in another function.)

---

> I think it's better to do something like

I think this `approx_P[~np.isfinite[M_T]] = np.inf` does not work because it doesn't change the distance profile of each index. In other words, the cleaned version `T` (after preprocessing), already affected the dist-to-NN of other `finite` indices. 

Minimum Working Example:
```
T = np.array([0,0,0,0,4,5,6,7,8,9,10,11], dtype=np.float64)
T[5:8] = np.inf
```
Then you can see the difference between `mp_correct = stumpy.stump(T, 3)`
and:
```
T_clean, T_sliding_mu, T_sliding_std = core.preprocess(T, 3)

mp_incorrect = stumpy.stump(T_clean, 3)
P_incorrect=mp_incorrect[:,0]
P_incorrect[~np.isfinite(T_sliding_mu)] = np.inf
```

How about just to avoid `prepocess(T, m)`? and pass raw `T` if possible? (I can check and see how it changes the design)

Reply:
<p>No, I don't think it's a good idea. I think we should just ust <code>_prescrump</code>  instead of <code>prescrump</code>.  I was trying to avoid it but I since we've already done <code>preprocess(T, m)</code>  then we should really be using <code>_prescrump</code>  as those distances will be correct since you are passing in the relevant <code>M_T</code>  <span class="ql-cursor">﻿</span></p>

Reply:
<blockquote>What do you think? In your opinion, is it a poor (or NOT a good) design? (because you had to notice another thing in another function.)</blockquote><p>I think it's fine. Let's piece it together and then I can see how it feels </p>

Reply:
> To some extent, I think about the private functions as the ones where the data has been preprocessed already while the public functions are there to catch and clean bad inputs.

I see it now. please allow me sometime to apply changes on my end (so, If that's okay with you, I will keep this conversation open here for now till I handle this)

Reply:
Sounds good! @Ninimama  you're quickly becoming a pro at this thanks to your open mindedness and hard work! I really appreciate you taking the time to contribute to STUMPY!

Reply:
I used `_prescrump()` (see below) and things are ok!

```
#calculate P

excl_zone = int(np.ceil(m / config.STUMPY_EXCL_ZONE_DENOM))

n = T.shape[0]
k = n - m + 1
P_squared = np.full(k, np.inf, dtype=np.float64)
I = np.full(k, -1, dtype=np.int64)

squared_distance_profile = np.empty(n - m + 1, dtype=np.float64)

s = int(0.001 * n)

for i in np.random.permutation(range(0, k, s)).astype(np.int64):
    QT = core.sliding_dot_product(T[i : i + m], T)
    _prescrump(
        T,
        T,
        m,
        M_T,
        Σ_T,
        M_T,
        Σ_T,
        QT,
        i,
        s,
        squared_distance_profile,
        P_squared,
        I,
        excl_zone,
    )

P = np.sqrt(P_squared)
```
and I may use a wrapper to encapsulate it in a function...

**Then, I was thinking about what you said:**

> I think about the private functions as the ones where the data has been preprocessed already while the public functions are there to catch and clean bad inputs.

So, what if I add a new input `T_isfinite` (where `T_isfinite = np.isfinite(T)`) in private functions? (and this is different than `np.isfinite(M_T)`).  

Then, instead of using the block above in function `_murlin()`, I can do:
`prescrump(np.insert(T_A, np.flatnonzero(~T_isfinite), np.inf) , m)`.

Yes, `prescrump` needs to  calculate `M_T` and `Σ_T` again but that should be a fast computation (and we are just doing it one more time in `_murlin` before calling other functions) 

---

So, I was wondering if you have any opinion. Do you think adding the block of code (or its wrapper) is still safer and having better design rather than using `prescrump`? (If it hard for you to see things now, I can do changes and push commits and then see how it goes)

Reply:
How do you feel about the new function below? (It is just a part of `prescrump`) 

```
def _self_prescump(T, m, M_T, Σ_T, s=None):
    excl_zone = int(np.ceil(m / config.STUMPY_EXCL_ZONE_DENOM))
    
    n = T.shape[0]
    k = n - m + 1
    
    P_squared = np.full(k, np.inf, dtype=np.float64)
    I = np.full(k, -1, dtype=np.int64)
    squared_distance_profile = np.empty(k, dtype=np.float64)

    if s is None:  # pragma: no cover
        s = excl_zone

    for i in np.random.permutation(range(0, k, s)).astype(np.int64):
        QT = core.sliding_dot_product(T[i : i + m], T)
        _prescrump(
            T,
            T,
            m,
            M_T,
            Σ_T,
            M_T,
            Σ_T,
            QT,
            i,
            s,
            squared_distance_profile,
            P_squared,
            I,
            excl_zone,
        )

    P = np.sqrt(P_squared)
    
    return P, I
```

I used the term `self`  to show that it is a `self-joins matrix profile`. Another option can be: `_premurlin` (but this name doesn't show its purpose) Although we do not need the matrix profile indices `I`, I returned it anyway to be consistent with `prescrump`)

Reply:
<p>I think I would just call it _get_approx_P(T, m, M_T, Σ_T, s)  and require <code>s</code>  and only return <code>P</code>  at the end (without <code>I</code>). </p>

Reply:
Thanks for the suggestion! All good!